aip/general/0129.md

```diff
@@ -0,0 +1,119 @@
+allocation.
+
+Some examples of these types of fields are ones that are:
+
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
+
+An attribute with an effective value ***must*** be expressed as two fields in
+the API:
+* a mutable field that can be optionally set by the user and **must not** be
+modified by the service
+* an `OUTPUT_ONLY` field that records the effective value decided on by the
+service
+
+Example:
+```proto
+message VirtualMachine {
+  …
+  string ip_address = 4;
+  string effective_ip_address = 5 [
+    (google.api.field_behavior) = OUTPUT_ONLY
+  ];
+}
+```
+
+#### Naming
+
+Effective values *should* be named by prefixing `effective_` to the mutable
+field's name. Consistent naming is important for identifying standard behavior
+across APIs and fields.
+
+### Client-Settable Fields and Normalization
+
+For client-settable fields, the value in response from the service **must** be
+the same as provided by the create or update request. For string fields this
+means returning the unchanged character sequence, with two exceptions: 
+
+* When a string field has a data type annotation, a normalized string that
+represents the given value may be returned.
+* When a string field has a default value annotation, that default value may be
+returned when the given value is empty or missing.
+Normalizations are important to allow services to store and return values in a
+standard way while communicating to clients what changes are semantically
+identical. Normalizing a value on the service side allows the service to accept
+a wider range of semantically identical inputs without needing to maintain every
+value as a raw string. Surfacing the normalization that is being applied to
+clients allows for client side comparison of sent and retrieved values to check
+for differences.
+
+For example, in a resource that accepts an email address on a particular field
+a client may specify a given email address in a variety of ways. For the email
+`ada@example.com` a client may choose to specify `ADA@example.com`,
+`aDa@example.com`, or `AdA@example.com`. These are semantically identical and
+*should* all be accepted by the service. The service then may choose to
+normalize the email address for storage and retrieval through downcasing or
+canonicalization. Importantly, the information surfaced to clients on the
+normalization of a field must not describe the normalization algorithm itself,
+but instead the comparison method used to accurately compute if two values
+should be considered equal.
+
+#### Allowed Normalizations
+
+This list of allowed normalizations is defined as follows:
+
+* uuid
+* ipv4
+* ipv6
+* email
+
+Normalizations on fields **must** be described using the `google.api.value_type`
+annotation.
+
+#### Field Value Handling
+
+For all values that do not have an allowed normalization applied
+[state-driven clients](state-driven-clients) will not be able to identify which
```

Comment:
we don't need to mention state-driven-clients here. I'd suggest moving a lot of the discussion around state-driven-clients to a `rationale` section, so the guidance is easy to read.

Those who want to understand *why* they need to adhere to this guidance can go to the rationale section.

Reply:
Done